### ISSUE TITLE
Cars Pulling Over for Emergency Vehicles

### ISSUE THREAD

**Issue by FireController1847** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#64]
As found in many countries, vehicles must pull to the side of the road and stop if there is an emergency vehicle behind them or coming toward them. If at an intersection, all vehicles must STOP at the intersection regardless of the light color until the emergency vehicles has exited. Reckless drivers could possibly ignore this rule entirely (as seen in real life, people can refuse to pull over. My parents always honk at those guys lol).

I plan on implementing this, but I must research a lot more into how to do it. Mostly making this issue for logging sakes and so nobody steals it ;)

**Comment by krzychu124** [comment]
Hahaha I'm sure nobody wants to steal this task

**Comment by krzychu124** [comment]
Which milestone do you want ? 1.10.16?

**Comment by FireController1847** [comment]
Heh, I don't know if I'd be able to get it done in time for 1.10.16. I think we should hold off the milestone for now and maybe 1.10.17. I can confirm to you if I can get it done for 1.10.16 when I make the PR xD

**Comment by krzychu124** [comment]
Just joking, great, feel free to ask for help, maybe I would find something while digging in other places 😉

**Comment by pcfantasy** [comment]
This is very very helpful in the game.

**Comment by FireController1847** [comment]
@krzychu124 I do have a question and one thing for you to look for, if you can.

1. What is a "Simulation Step" and how does it work?
2. If you can find out how to get cars within a segment, that'd be awesome. I think I asked about this in Turn on Red, but didn't need it to do what was tasked. Now, to do this, I'd need to detect if there's an emergency vehicle in a nearby segment, which means I need to get the cars from within a certain segment.

**Comment by pcfantasy** [comment]
@FireController1847

	uint num2 = vehicleData.m_path;
	byte b = vehicleData.m_pathPositionIndex;
	byte b2 = vehicleData.m_lastPathOffset;

	PathUnit.Position position;
	instance.m_pathUnits.m_buffer[(int)((UIntPtr)num2)].GetPosition(b >> 1, out position)

       You can find vehicle is in Position.m_segment.  You need a for {} to find cars within a certain segment

**Comment by FireController1847** [comment]
@pcfantasy What object is that in? PathManager, if there's such a thing? Can I know the filenames for where you found those?

**Comment by pcfantasy** [comment]
@FireController1847

VehicleAI.UpdatePathTargetPositions

In fact, those data can be read in everywhere, For example

You can do it in  onBeforeSimulationFrame();

for (i = 0; i< 16384, i++)
{
   vehicle vehicleData = Singleton<VehicleManager>.instance.m_vehicles.m_buffer[i];
   var instance = Singleton<PathManager>.instance;
uint num2 = vehicleData.m_path;
byte b = vehicleData.m_pathPositionIndex;
byte b2 = vehicleData.m_lastPathOffset;

PathUnit.Position position;
instance.m_pathUnits.m_buffer[(int)((UIntPtr)num2)].GetPosition(b >> 1, out position)
//then you can get everything use my above codes
}

**Comment by FireController1847** [comment]
@pcfantasy I hate to ask of this, but do you also know how to get the targetNodeId? I need to detect whether or not the vehicle's target is the intersection, or if the vehicle has left the intersection and everything can return to normal.

**Comment by FireController1847** [comment]
Okay, now I need some opinions because there's a few ways I can do this. *I'm currently implementing the "stop at a light no matter what if oncoming," not the pulling over aspect.*

1. The first way is that I can use the VehicleBehaviorManager to check and see if there's an emergency vehicles within the opposing segments, and if their targets are the intersection. This would not only require a loop and filtering of all vehicles to find the ones on the segments, it has to happen for every segment connected to a node, AND it has to recalculate the position of the emergency vehicle.
2. I can add an attribute to segments with traffic lights like "HasEmergencyVehicleApproaching." Then, I can use the VehicleBehaviorManager to handle accordingly. This would require a modification of the emergency vehicle AI to where if the target node has traffic lights to update that node that it's approaching, and then when that node is no longer its target to set the boolean back to zero.

These are my two suggestions on how to do this. Do you guys have any ideas? I've been attempting the first one without realizing what a huge task it is, so I think I want to go with the second one. But if you guys can come up with better ideas, then please let me know.

Also, do any of you guys have Discord? If so, would you mind if we chatted about it? I'm `FireController1847#3577`

**Comment by pcfantasy** [comment]
@FireController1847

//you can try blow codes, I have not test this.

targetNodeId = Singleton<NetManager>.instance.m_segments.m_buffer[(int)this.SegmentId].m_endNode

//maybe there will be a reverse
targetNodeId = Singleton<NetManager>.instance.m_segments.m_buffer[(int)this.SegmentId].m_startNode


//So there must be another vehicle direction to decide which is the correct targetNodeId

**Comment by aubergine10** [comment]
Why not add an ```EmergencyWeight``` integer to segments?

Emergency vehicles then increment the value a certain number of segments ahead (in diagram below, 4 segments, but could be any number depending on what works best) and decrements the number behind it.

```
      4     3     2     1
V -> ____  ____  ____  ____
```

So the emergency vehicle would create sort of a wave or ripple ahead of itself, warning segments that it's on the way.

If multiple emergency vehicles are heading the same way (common with fire engines) then that would create multiple overlapping waves, ensuring 'moved to the side' vehicles know to wait before continuing their journey.

Other vehicle AIs would check the ```EmergencyWeight``` for the segment they are on and take action accordingly. Likewise, traffic lights would also take account of the value. Even toll booths could prepare themselves by allowing a lane of traffic to pass without paying to ensure the emergency vehicle has clear route. Busses could avoid pulling out of bus stops... anything that's travelling on or part of the road network can adapt by looking at the ```EmergencyWeight``` value.

Note: I'm hopeless at monomorphic coding (like C#) so if what I've said above is stupid don't hesitate to ignore it (I know enough to know that I don't know enough!).

**Comment by krzychu124** [comment]
Generally I like your idea 👍 , but I think it would look quite strange when vehicles were started to pulling over up to 3 junctions ahead of ambulance (imagine that ambulance has to follow this route:
```
                * <- target
                |
        *-------*
        |
V ------*
```

**Comment by aubergine10** [comment]
They only need to start pulling over when it the weight reaches a certain value (for example, 3 or 4).

But other things might need to adapt sooner, such as traffic lights and toll booths (they would start acting when the weight is lower, for example 1 or 2).

There are some problems with the general approach I propose, for example what if an emergency vehicle despawns for some reason (eg. user uses "Clear Traffic" tool)? So what it might need to do is at the start of each frame clear all the weight values; that way emergency vehicles don't need to worry about clearing up after themselves so much.

Furthermore, remember that once a vehicle passes, if there are others nearby on same route, you don't want "pulled over vehicles" to start driving back in to the road, they should wait for 'weight = 0' before doing that.

**Comment by Strdate** [comment]
I would say that the important part of this code is to know when to trigger the vehicle pathfinding recalculation. I don't know if there is a list of vehicles on current nodes, but it should be because how would otherwise traffic lights and stuff like that work? I recommend looking at their source code.
It think that for this to work well you would need to allow the vehicle to use the opposite direction lanes.  Overall, this feature is one of the most difficult tasks. Then there is the limitation that vehicles cannot switch lanes anywhere else than on junctions (taking into consideration the virtual ones as well). (Or am I wrong?) So it is a question how realistic the behavior will be.

**Comment by aubergine10** [comment]
When vehicles park, or when things like ambulances and hearses pull on to pavement, the code that does those things would be worth inspecting.

IMO it would be worth doing some mini mods, like on that just deals with making vehicles on a segment pull over when segment is clicked, and then resume driving when segment clicked a second time. People could test that out and report back any obvious bugs or issues. I think if we try and release this feature in one big update, there will be so many bugs that we won't know where to look.

**Comment by Strdate** [comment]
I agree with this. I think that slowing down is ok, but anything more complex like going on the side is something which requires more than what usually happens in TMPE mod.

**Comment by krzychu124** [comment]
AFAIK ambulances and hearses pull on to pavement because every building in game has set spawn point there which in this case is used as destination target

**Comment by aubergine10** [comment]
When a road is demolished and new one built nearby, cars will travel from their current location to the new road (if it's near enough) - so we could simply 'block' (somehow) the affected lanes and repath nearby vehicles (they might do it automatically without us needing to repath them?) - they will swerve on to remaining paths.

**Comment by FireController1847** [comment]
While yes I agree this is a very complicated feature with many ifs and buts, I don't think it's too much to add to TMPE. TMPE is supposed to give people control over their traffic, right? By allowing people to have cars pull over for emergency vehicles, it gives even more control over their traffic. Overall, though, I'll let @krzychu124 decide if I should make this a separate mod or not.

On another note, @aubergine10 I like your idea of having a "weight" for emergency vehicles. However there's one major issue: Segments are *huge*. If we were to do 3-4 segments down, it would be the equivalent of stopping vehicles miles away. Now I know you said they'd only stop on the highest one, and I think that's a great idea, but that provides us with a problem of needing to calculate three to four segments in advance for the emergency vehicles. I'm having mixed feelings about this but I feel this is our best option, honestly.

When it comes to traffic lights, we will need a way of detecting if there's a blocked path. When in a traffic jam, vehicles cannot pull over or cannot move. Due to this, vehicles must continue moving forward as if there wasn't an emergency vehicle there until there is a clear path to the intersection. I have a great example of a blocked road in one of my maps at home, so later today I'll be able to send an image of it.

Having vehicles pull over and stop does not seem very difficult. Would it not just be setting their trajectory like how busses pull over, like mentioned? Busses do the exact thing that cars need to do.

I agree that this is a daunting task, but I believe you guys are making it more complicated than it needs to be really. There's three major things we need to do for emergency vehicles:
- Pull over.
  - To implement, we'd need to add a way to get the vehicles to pull over. A toggle test button within their menu, perhaps.
  - We'd also need a way to detect when there's an emergency vehicle *approaching*.
    - It wouldn't make much sense to pull over if it's not approaching them from either direction, no?
  - To detect is not that difficult, but there's different ways we can do it.
    - The recommended way is a weight system, which sounds complicated, but makes sense for future implementation. It wouldn't be too difficult, either.
    - We just need to figure out the first three segments ahead and the last three segments behind.
    - This would result in us having to speed up pathfinding or get it further ahead, which is certainly possible and not unrealistic.
    - We will definitely need help with this feature, and need an understanding of the pathfinding.
- Stop at traffic lights.
    - This will also require detection of vehicles.
    - We need the ability to resume traffic if the vehicle is stuck behind them.
- Resume normal traffic.
As seen here, there are many if's and but's when it comes to this feature, and yes there's a lot of test cases, but if we just follow these general three rules without attempting to add more features like toll booths and other things, it gives us a basis to work on. Then we can _possibly_ add features like toll booths.

Don't look at this from a general perspective (unless you're brainstorming about the entire feature, in which case feel free), but look at this from the initial three points. If we add features slowly, step by step, instead of trying to tackle it all at once, we'll actually be able to do something.

**Comment by Strdate** [comment]
Nobody wants to exclude this from traffic manager, but the idea is that this requires testing of so many features which should be rather done outside TMPE. This is indeed very daunting task, it is connected to half of other features we would like to implement.

**Comment by aubergine10** [comment]
As @Strdate says above, community testing is going too be important for this feature - it has to be tested on all sorts of roads, junctions, etc. What effect does it have on trams, bikes, busses? What about roads with multiple medians and weird lane arrangements... The end result will be in TMPE, but initial dev might be better suited to a separate mod where we can easily roll out changes to a large audience of beta testers rather than waiting for TMPE release cycle.

Anyway, for the cars pulling over, I think the best place to look might be how parking lanes work. That's almost exactly what we want (more so than bus stops), we basically make an entire lane of traffic "park". We could perhaps create fake parking lane for them to park in or something? That way we can reuse vanilla code that does the parking manoeuvres. 

If we release a mod with that "pull over" feature, activated by clicking a segment to toggle between "pull over" and "go back to normal", it will allow lots of people to test it out and let us know what happens.

**Comment by krzychu124** [comment]
The best would be create separate github repositories and maybe Steam workshop items to test small parts of this feature (As @aubergine10  there would be tenths of different roads types to test with). 

Later we could try to merge those small parts to create something bigger and test again.

I've been thinking about splitting TE:PE to be more modular and open on extensions (plugins) like other great big Steam workshop mods. Tell me what do you think. Is it possible and worth effort?

This feature could be a sort of plugin to mod base, but first I need to make some tests to check if it is possible.

**Comment by FireController1847** [comment]
@krzychu124 I don't think it's worth the effort. I know a lot of people would rather have a mod that can be installed with one click and then modified how they wish, versus having to try and find a bunch of different workshop extensions which may or may not be updated and may or may not be listed in the description which may or may not be working. When it's all in one big bundle, you can be sure it will either work or it wont -- or at least, some features will work and some wont. A great example of bundled would be the NeXT mod, as well as how TMPE is now. You can see the total downloads and are sure that it will either work or not. Now look at the mess of a mod called Roads United. There's many different versions of the base mod, some of which work and others don't, there's many extensions that sometimes work and sometimes don't, and there's often confusion between which mods are required and which mods should be enabled. I personally would much rather have a mod like TMPE is now where you can enable and disable settings as you please, all in one place.

@aubergine10 @Strdate Apologies of my misunderstanding. I think it's a good idea to make a separate workshop mod & GitHub repository to test these features, however I would have no idea where to start. How do I access TMPE methods? How do I set up Visual Studio? Do I make it a completely separate mod? Also, what if I need to use protected/private methods? What's the plan there?

**Comment by krzychu124** [comment]
With regards to setting up Visual Studio and accessing methods, just add compiled TrafficManager.dll to project references and you will have access to everything (public). Reflection should help with protected/private methods - it's just test it doesn't have to have max performance, I think.

**Comment by aubergine10** [comment]
@krzychu124 I agree with @FireController1847 , I think TMPE is best as single mod for now.

Even back in early T++ and TM days, back in 2015, people were asking for modularisation. But it became clear that doing so actually caused performance issues. The approach TMPE takes currently seems to work well in that it recompiles itself with just the activated features, thus completely removing unnecessary code from runtime. Imagine if it were a bundle of separate mods, all having to interact with each other to some degree. Also, what if in a year or so, we've all gone off to do other things and someone else takes over with a new branch of the mod. I think this would be confusing for end-users to try and work out what component mods work with the main mod; like @FireController1847 says, we'd end up with a "roads (dis)united" situation.

I can obviously see the benefits of modularisation, however, such as being able to rapidly iterate individual features without needing to roll out the full mod, but in practice I think it always turns out not to be as simple as that. Another example is the various 'extensions' to "Move It" mod that were released; despite the main Move It mod being stable and not changing much any more, those extensions were short lived and in the end it was easier for quboid (sp?) to just release a new Move It with all those features bundled in to the  main mod. Compared to TMPE, Move It is a simple mod, so...

**Comment by krzychu124** [comment]
Thanks @FireController1847 and @aubergine10 for your opiions 👍 So now we can focus on more important things 😉

**Comment by FireController1847** [comment]
I think I'm not fully understanding this "separation" thing. How do I make a mod that modifies the logic of TM;PE? Particularly, the driving logic? I would need to interrupt the normal driving logic (which, by the way, I have no idea where it is) with the parking logic. Wouldn't that require me to make a modification to the AI, which you can't do if you only extend/use functions within TMPE?

**Comment by krzychu124** [comment]
In this mod we have pathfinding function which finds paths from source to destinations and vehicle behavior which is responsible for their travel along path (stop, move, wait etc.) In case we want to force vehicles to pull over I think vehicle behavior class would be better place to start

Just throwing my ideas below... 

I suggest to digging through code to find what is going on when ambulances driving on emergency (Flag.Emergency2 in vanilla). 
Sometimes I can see that ambulances are moving in-between lanes or turning right with overtaking other vehicles at junctions. 
There must be a line in code where their path bend is performed.

**Comment by FireController1847** [comment]
Alright. I'll look at some of that path stuff this weekend.

**Comment by FireController1847** [comment]
Okay, I've began the creation of the separate mod. You can find it at [FireController1847/Cities-Skylines-Emergency-Vehicle-TMPE-Extension](https://github.com/FireController1847/Cities-Skylines-Emergency-Vehicle-TMPE-Extension). So far I've only added the button to the vehicles, not any cool things like the beginnings to pulling over. Tomorrow I will research more into that :)

**Comment by FireController1847** [comment]
So I've been experimenting (this is like my favorite phrase now I guess xD), and I've found some things. First, I found that "FindBestLane", which is part of VehicleAI, it used during the SimulationStep. TM:PE does not currently override this method, so I did using the mod. And it works, however I think I may be in the wrong area.

If we want to detect if there's an emergency vehicle, and have the vehicles pull over, should I be using SimulationStep, or should I be using something else? Because if I'm in the fundamentally wrong area, then I don't want to waste hours on something that wont work.
Using ILSpy, it's clear that the FindBestLane thing relates directly to the parking of vehicles within the simulation step. Within TM:PE's custom simulation step for vehicles, there's a comment called "park vehicle." Does this have to do with anything, or does it not do what I need it to?

**Comment by krzychu124** [comment]
It's fun to watch you fighting with game codebase (edit log) 😄 

There is a thing which could greatly help you with investigations and testing. 
It's  _magic_ tab with name __Call Stack__ if you are debugging with dnSpy 😃 
* just put breakpoint inside code e.g. path finding method
* go to that tab or open if it's not
* now you should see _call stack of methods_

In game we have a lot of threads but we can separate them on some general:
* one main for rendering graphic (not sure)
* one for simulations (calls simulation steps of every *AI class)
* up to 4 pathfind threads _(line ~19 original PathManager.Awake())_
```int num2 = Mathf.Clamp(SystemInfo.processorCount / 2, 1, 4);```
* one for audio 
* and more for other stuff (water simulation, file monitor, mods compiler, mesh simplification, background workers => CPU cores *2 )

No worries I will try to help you later today.

**Comment by FireController1847** [comment]
@krzychu124 Sorry, I'm not normally a C# coder (I'm more Java), so I'm still learning all of this visual studio stuff >.< And just _wow_, dnSpy is amazingly useful. Thanks for suggesting it to me.

I'm unsure where to find the path finding methods. Is that were we should be putting this? I'd assume so, that does make sense (if there's vehicle pathfind to tell card to pull over and step).

Sorry I wasn't able to "be helped" throughout the day, I think we may be in very different timezones (as of this post it's 3:44 PM for me), but I try my best to be as active as possible 🙂

**Comment by krzychu124** [comment]
No problem I am Java coder too (and Typescript/Javascript). C# was my first language that I've learnt and I still like it.

> I'm unsure where to find the path finding methods. Is that were we should be putting this? I'd assume so, that does make sense (if there's vehicle pathfind to tell card to pull over and step).

Me too, and that pathfind sorcery... I'm still trying to understand what is going on there, so for now I can't help here. Give me a few days...

```CustomPathFind2``` is current pathfind implementation and ```CustomPathFind2``` is older - currently not used.

Haha timezones. I see 11:56PM on my system clock.

**Comment by FireController1847** [comment]
Alright, not a problem. I'll also take a look and see if I can't understand some of it too! 🙂

**Comment by krzychu124** [comment]
It's hard to read if you see those preprocessor directives (for conditional compilation) everywhere...
```cs
#if CONFIGURATION_VARIABLE
#endIf
```

**Comment by aubergine10** [comment]
I'm not sure it's the pathfinder you want; we're not changing routes of vehicles, just asking them to temporarily pull over. So, IMO, it should be in the vehicle AIs - they pause from their path to let emergency vehicle get through, then return to their path. That being said, I also am still wading through code trying to work out how all this stuff works so I could be talking rubbish.

**Comment by FireController1847** [comment]
@aubergine10 No, that appears to make a lot of sense. I've been looking more in-depth at the CarAI.SimulationStep (with 5 arguments), which appears to be the area we want. I got here by following the calls to the VehicleBehaviorManager (which stops cars for junctions). My current goal is to literally just add a button that makes cars stop.

**Comment by FireController1847** [comment]
Huzzah! I've reached my goal of making all the cars stop. After wondering why my override went into an infinite loop (crashing the game), I finally got it to work. This is the first big step!

To test it, build & enable my "Emergency Vehicle Extension" mod, and then enable the "buildings" overlay for TM:PE. When you enable it, all cars stop moving. When it's disabled, they resume.

**Comment by aubergine10** [comment]
I was looking at code and noticed that there are separate simulation steps for things like flooded (car washed away) or blown (car sucked in to tornado)... Were you planning to do something similar with the stopping for emergency vehicles? As separate simulation step could make them stop or pull over or something?

**Comment by aubergine10** [comment]
Link to repo: https://github.com/FireController1847/Cities-Skylines-Emergency-Vehicle-TMPE-Extension

**Comment by FireController1847** [comment]
That's actually not a bad idea, I'll check into that at some point this weekend or this week (probably this weekend, though).

On another note, @krzychu124 @aubergine10, do you guys think it's a good idea to add the ability to get the cars from a lane? I think it could be useful for future issues (even though this does not pertain to this issue), especially if we need to fetch the vehicles around a vehicle. I can't think of many other use cases for that, though.

**Comment by aubergine10** [comment]
We'll probably need to get cars from a lane anyway... We'd have to decide a lane for the emergency vehicle, and then move cars either side (where applicable) of that lane out of the way so the emergency vehicle has a clear lane.

As for adding vehicle flags, I have no idea how to do that, however I would suggest investigating the code of the Real Time mod: https://github.com/dymanoid/RealTime and ideally chat to Dymanoid. From what I understand, he created some sort of 'extension' for citizens so he could detail their daily routine, and I imagine the same approach could be used for vehicles (and transport networks too I suppose).

Dymanoid really nailed the development of that Real Time mod - not only did he add loads of stuff in a very reliable manner, it barely affected performance of the game. I think there's likely lots of his coding techniques that would benefit TMPE.

**Comment by aubergine10** [comment]
saving these for prosperity

![garbage_truck](https://user-images.githubusercontent.com/1386719/53532975-e4ff2800-3af0-11e9-93f0-50a983c1c212.jpg)

![giphy](https://user-images.githubusercontent.com/1386719/53532978-e6c8eb80-3af0-11e9-872e-34d6075e30c8.gif)

**Comment by aubergine10** [comment]
![bus](https://user-images.githubusercontent.com/1386719/53640075-03654080-3c23-11e9-947a-1c39b7e7e095.png)
![car](https://user-images.githubusercontent.com/1386719/53640076-03654080-3c23-11e9-8414-fffcfafa1e14.png)

**Comment by Strdate** [comment]
From the screenshots it looks soo good!
Maybe garbage trucks travelling in the opposite direction is a bit of an overkill (it would be too easy for the player and it is not too realistic), but emergency vehicles can really use that!
Pls update the github repo if you have time! 🥇

**Comment by FireController1847** [comment]
@Strdate Those were test cases that were modified live using the debugger, there's not any actual in-game code to run this. However, I am currently implementing that, while @krzychu124 is currently looking at ways to get the vehicles to pull over! :)

**Comment by aubergine10** [comment]
@FireController1847 Can you upload that video of the  police car going crazy to youtube or something for easier sharing; we have to save that for prosperity!

**Comment by VictorPhilipp** [comment]
Last year I have started to experiment with implementing this feature. I will post some videos and the code as soon as I get the old code up and running.

**Comment by krzychu124** [comment]
Great!

**Comment by FireController1847** [comment]
Ask and you shall have!

https://www.youtube.com/watch?v=HWcx82qaJbs

**Comment by ntoff** [comment]
The cities skylines subreddit would go nuts over that video, post it and title the thread "when the donuts are on sale", you'll be the hero cities skylines needs.

In terms of this
>The best would be create separate github repositories and maybe Steam workshop items to test small parts of this feature

Just create a new branch here on github, no need for separate repositories, and I'd definitely be down for a "TM:PE[testing]" branch/mod on the steam workshop (similar to the way the loading screen mod has the stable and test versions). I'm currently am running a self-compiled version of TM:PE that I re-build when I see the code here changes, so I'd be down for some testing.

**Comment by aubergine10** [comment]
> Here where I live

Which side of the road do vehicles drive on where you live? Sounds like left-hand traffic?

In UK, cars will generally pull over to the outer edges of the road, freeing up space in the middle. On motorways, however, there is a dedicated emergency hard shoulder on the outside (where broken-down vehicles wait for recovery, etc). See also: #26

**Comment by ntoff** [comment]
Yes, we drive on the left. We do have that hard shoulder, but it isn't consistent, it's "better" if the emergency vehicles just travel up the fast lane as it has no on or off ramps attached to it, so no chance of them collecting an idiot merging who isn't watching the traffic. You'd be amazed at how many people here try to merge onto a 100km/h highway at 40km/h and THEN accelerate to 100km/h.

**Comment by FireController1847** [comment]
@ntoff I think you're misunderstanding the point of the separate repository; it's for testing. If we made a different branch, since I'm "in charge" of the feature, I'd have to constantly make PR's to update the branch. With a separate repository, I can update it at any point in time. Also, with a separate repository, I can make it act like a separate mod, meaning I could publish this mod to the workshop so users can officially test it before it's implemented. Once we're ready to actually implement it, then we'll probably make a PR similar to how we did with #25.

Now, for everyone else, prepare for a long essay...

Emergency vehicle logic currently attempts to find the lane with the most clear traffic, basically meaning they take the route of least resistance. To have vehicles pull over takes majority in most countries, including the U.S., most of Europe, Japan, and similar places. "Pulling over," to make it easy to implement, refers to pulling to the preferred side of the road where the gutter is (right-hand to the right, and left-hand to the left) and stopping. I am fully aware of many places that make a separate "center" area that is clear, but that is in few select places and is not the majority.

You have to remember, around the world there's hundreds, if not thousands of different ways traffic works. If we were to implement every way, it would take us years to even get started. So instead, we need to choose the most popular form of road management - America and Europe. Yes, we have users in Korea. Yes, we have users in Japan. Yes, we have users in Russia. But is that _all_ the users? Is that even the majority? So when we implement this, we will most likely base it off of the basic logic: Pulling to the right (or left) to get out of the way, stopping at intersections, and moving out of an emergency vehicle's way. And if the other users don't like it, there's an amazing thing called _options_ where they can disable it.

In the future, it would be awesome to add more test cases and unique opportunities, but as it stands it's _already_ super complicated, and has many many "ifs" and "buts." We need to implement the basics first, which is what we're currently testing. So hold your horses everyone, we're working on it. This is a massive feature and @pcfantasy was correct, it was a much larger task than we originally anticipated.

In terms of if this is even a good idea or should be added to TM:PE, it definitely should. As @VictorPhilipp stated in our development Discord channel, it is going to be complicated and definitely will deteriorate performance. But, what doesn't the first time it's implemented? There is no reason not to test it, add it, and see how it works. If we said every feature would deteriorate performance, then we wouldn't have a "smart AI." We wouldn't have custom traffic lights, and we definitely wouldn't have priority rules. It will take time, effort, and lots and lots of testing, but we will be able to get this implemented efficiently. There's no reason to give up.

Nothing is impossible. If I have to take a completely roundabout route to get the vehicle behind a selected vehicle, what's to say I can't simplify that roundabout route? I may have to get into the nitty gritty of 3D space, XYZ directions, and bezier curves, but if that's what I have to do then so be it. Unity is a powerful engine, and it is NOT impossible to get objects in 3D space in relation to those other objects. And, not only that, but possibly by adding the simplicity of fetching vehicles in relation to things like lanes and segments, we could add more features that could be dependent on that logic. Possibly even helping vehicles make decisions based on the position of other vehicles? There's a LOT of power in that type of logic.

On a different note, in terms of TM:PE as a whole I believe that it is more than a "Traffic Manager" at this point. It is probably more along the lines of an "Advanced Traffic" mod, which adds features to traffic to make it more realistic. If you google anything about roads or intersections, Cities: Skylines is one of the first things to come up for examples. C:S is one of the best simulations of traffic that games have seen, and TM:PE just adds to that realism. Since TM:PE is supposed to improve traffic, there's no reason not to add more features to make it more realistic. It should be _expected_ to deteriorate performance (not that C:S is a very performance-based game... it already sucks in terms of performance). And when options are disabled, performance gains can be found.

The only reason I'm making this comment is because there appears to be a lot of doubt about this feature, and I want to put a lot of that to rest. Even if everyone else gives up on this feature, _I won't_, I'm very adamant about doing this. So whether or not this gets kicked off of the TM:PE future, I'd still most likely try to add it in a different mod.

Now, to make decisions about the feature...

I think that the feature should be based off of North American logic for the majority of its development. _It is not a bad thing to add country-based features_. If we add the North American logic of Emergency Vehicles to the feature, we can then modify the already implemented logic to fit with other countries' standards.

The main three points of North American logic are the following:
- Vehicles must pull over and stop for Emergency Vehicles.
  - Once they have passed, vehicles may resume normally (they can even follow behind EV's, as long as they don't pass them).
- If unable to pull over, vehicles should try their best to get out of the way, or get into a situation where they can pull over.
  - This means that if there's an EV behind you, and you're stuck at a red light or in a traffic jam with no space to pull over, you should either wait for the light to turn green, cross, then pull over, or turn right and pull over, then make a u-turn or whatever to resume where you were going.
- If you know there's an EV approaching an intersection, you must stop and give the EV full priority, regardless of signs, light color, or anything similar.

The easiest to implement would most likely be the first one. Then we can begin the other two points, starting with the third 'cause that'd be the second easiest. There's also some improvements we could make to the EV logic:

- Emergency Vehicles are allowed to drive in the opposite lane of traffic _if it's safe_.
- EV's should slow down a LOT when approaching an intersection to avoid a crash (they almost stop sometimes).

These two are pretty easy to do, and as you can see in my previous video I've already completed partial logic for that.

So, I'd love to know everyone's opinions on the claims made in this comment. I'd also like to know opinions on the feature itself, and what guidelines we should follow. After all, we can't implement something if we don't know exactly what we're implementing.

**Comment by aubergine10** [comment]
The driving in opposite lane thing was truly epic win. I'd love to see small percentage of reckless drivers do that sort of stuff :) As for all the other stuff, just take it one step at a time.

As I understand it, the current issue is finding a decent way to get a list of traffic in a given segment? Is there anything in vanilla game that does similar, for example the code that identifies leader vehicles? Or the code that floods use to work out which cars get caught in the flood?

**Comment by aubergine10** [comment]
Pulled this image from #185 - some discussion in there that may be relevant to this thread.

![41798341-9a7adfe2-766d-11e8-9e21-29a6ed97fb96](https://user-images.githubusercontent.com/1386719/53691880-7931ef00-3d7e-11e9-9722-888873f2433f.png)

The police car video above shows we are already part way to achieving some of this.

**Comment by VictorPhilipp** [comment]
Status report for my EVE prototype:

[EVE feature.pdf](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/2923665/EVE.feature.pdf)

**Comment by aubergine10** [comment]
With the experimentations done by @FireController1847 I assume we'd need an additional mode: `FeelFreeToDriveInOppositeLaneAndDontBotherAboutThosePeskyMedians` :D <3

**Comment by FireController1847** [comment]
I want to make a few notes about this feature.

- Before I go any further, I plan on waiting for #260 to be completed, as it will make development of this feature much easier.
- I also wanted to say that the amount of emergency vehicles going around seems to be a little excessive within the game (however, I don't have much data to back that up), so possibly it might be work tweaking how often emergency events happen (especially fires). How often do you see a fire in your city? Every day? Every hour? And I mean roaring fires that will tear down buildings next to them. Honestly, that's way too high, so I don't know if this is within the scope of TM:PE but to make this feature happen it could seriously deteriorate traffic flow if the amount of emergencies actually happen stays at the rate they are.
- Hence, @VictorPhilipp, why I thought it was important to have vehicles immediately return back to the road instead of waiting until they're out of the segment. Segment-based EV detection is very good, but could quickly deteriorate traffic flow.

Man, the amount of logic to this feature is *PHEW*.

**Comment by aubergine10** [comment]
In terms of frequency:

* Vanilla game sends emergency vehicles all over the map, rather than keeping them local
    * Mods like District Service Limit 3 help a lot with that
* Fires: Very few (or maybe they get put out while I'm working on other part of map?)
    * However a fire will often spread fast and requires multiple fire departments to respond
* Police: Very common, especially as city grows and in low-density areas
* Ambulance: Common, especially as city grows
* Deathcare: Moderate usage, until death wave hits then it's like a hearse festival
* Disaster recovery: Only if something collapses (I think?) so rare until disaster hits

Note: I tend to reduce capacity (with AVO) of many of my vehicles so they make shorter trips (fewer buildings visited) but more of them.  I use Customize It to reduce number of vehicles per station/depot to add realism.

In terms of returning to road, if other cars on segment have pulled over are they not using less CPU time on the simulation step? "I'm waiting for this ambulance to go past, skip to next vehicle"

**Comment by FireController1847** [comment]
@aubergine10 What I meant about deteriorating traffic flow is not performance, I mean literally traffic flow itself. Cars would be waiting for something they don't need to be waiting for, therefore making more cars wait with them. As the EV goes around, it could quickly stop traffic across your entire city due to there being so many so frequently. Also, on major major roads, and main routes, an EV could be going back and forth on it very often. The more that road is stopped, the worse traffic there will be for every other road waiting to enter that road. Of course, this will add strategy to placing these emergency places, but without cars immediately returning back, it could quickly change traffic flow for the worse.

**Comment by aubergine10** [comment]
Ah, yes, very good point! In that case I agree it would be best to get cars back to their driving sooner rather than later (unless there are more EVs behind the one that just passed, something common with fire engines).

**Comment by aubergine10** [comment]
I wonder if it would make more sense to move the vehicle AIs in to the roads...

Like, instead of iterating vehicles and doing all those calcs (and leader car stuff, etc), to make the car drive along the road, why not make the road drive the car?

The road could then cache the routes along it on a segment-by-segment (lane-by-lane?) basis, so when a car enters the segment the vehicle motion is already pre-determined - then it's just a case of tweening along that motion path taking in to account vehicle speed. The same could apply to on-road parking, and also vehicles moving out of way of EVs, bits of that could be cached. There would be no need for leader cars any more.

**Comment by FireController1847** [comment]
@aubergine10 I like that idea tons. Like I've suggested a lot, there needs to be able to get a car's position relative to the road, which has been an impossibility recently without a lot of extra work. I like the idea of storing the data inside the road instead of the vehicle and having the road drive the car. I agree it would improve performance a lot! However it'd also require a _huge_ overhaul of the mod. Since I'm waiting for #260 anyways, maybe we could implement this alongside/after that?

**Comment by aubergine10** [comment]
> storing the data inside the road

Issue there is what happens when road bulldozed? We'd need a list of orphaned cars/cims and then work out what to do with them in subsequent frames.

**Comment by FireController1847** [comment]
Hello everyone! Apologies for my absence, however school is starting soon and this year is going to be one of my busiest yet. I have High School & College & A Pyrotechnics Class & A Club & My Personal Major Project. So, I'm going to have to pass this issue over to @VictorPhilipp in hopes he'll be able to continue his work. I didn't mean to take it over in the first place, and I clearly don't have the knowledge in C# to do it, so I hope you can take it where I've always envisioned it.

If it has yet to be implemented in the future if/when I come back to TM:PE (because I don't want to abandon it forever!), I'd love to implement it. If not, I'd love to help improve it or maintain it if necessary. Thank you everyone for your patience with my (very immature) emotional outbursts and your understanding of what's going on. I wish to see you all again soon!

**Comment by Leo512bit** [comment]
Shame that it seems that @VictorPhilipp is no longer active.
